Fix the timestamp field of a data stream to @timestamp

martijnvg opened the pull request:
The commit makes the following changes:
* The timestamp field of a data stream definition in a composable
  index template can only be set to `@timestamp`.
* Removed custom data stream timestamp field validation and reuse the validation from `TimestampFieldMapper` and
  instead only check that the _timestamp field mapping has been defined on a backing index of a data stream.
* Moved code that injects _timestamp meta field mapping from `MetadataCreateIndexService#applyCreateIndexRequestWithV2Template(...)` method
  to `MetadataIndexTemplateService#collectMappings(...)` method.
* Fixed a bug (#58956) that cases timestamp field validation to be performed
  for each template and instead of the final mappings that is created.

Relates to #58642
Relates to #53100
Closes #58956
Closes #58583

martijnvg committed:
Fix the timestamp field of a data stream to @timestamp

The commit makes the following changes:
* The timestamp field of a data stream definition in a composable
  index template can only be set to '@timestamp'.
* Removed custom data stream timestamp field validation and reuse the validation from `TimestampFieldMapper` and
  instead only check that the _timestamp field mapping has been defined on a backing index of a data stream.
* Moved code that injects _timestamp meta field mapping from `MetadataCreateIndexService#applyCreateIndexRequestWithV2Template58956(...)` method
  to `MetadataIndexTemplateService#collectMappings(...)` method.
* Fixed a bug (#58956) that cases timestamp field validation to be performed
  for each template and instead of the final mappings that is created.

Relates to #58642
Relates to #53100
Closes #58956
Closes #58583

martijnvg committed:
fixed tests

martijnvg commented:
@elasticmachine run elasticsearch-ci/2

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

martijnvg committed:
fixed more tests

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:
```diff
@@ -256,6 +257,10 @@ public String toString() {
         private final String timestampField;
 
         public DataStreamTemplate(String timestampField) {
+            if ("@timestamp".equals(timestampField) == false) {
```
Can we make this a public static string so it can be referenced elsewhere?

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:
```diff
@@ -267,6 +272,13 @@ public String getTimestampField() {
+         * @return a mapping snippet for a backing index with `_timestamp` meta field mapper properly configured.
+         */
+        public Map<String, Object> getDataSteamMappingSnippet() {
+            return Map.of("_doc", Map.of(TimestampFieldMapper.NAME, Map.of("path", timestampField)));
```
Super minor:

```suggestion
            return Map.of(MapperService.SINGLE_MAPPING_NAME, Map.of(TimestampFieldMapper.NAME, Map.of("path", timestampField)));
```

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -213,67 +207,37 @@ public int hashCode() {
-                "invalid type defined for mapping of timestamp_field";
 
+        public TimestampField(String name) {
+            assert "@timestamp".equals(name) : "unexpected timestamp field [" + name + "]";
```
I think we should transform this into an actual error, that way if we end up using it for the HLRC it doesn't fail to throw because assertions are disabled

martijnvg replied:
I think HLRC uses its own `DataStream` pojo. I will change this from an assertion to an actual error.

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -213,67 +207,37 @@ public int hashCode() {
         private static final ConstructingObjectParser<TimestampField, Void> PARSER = new ConstructingObjectParser<>(
             "timestamp_field",
-            args -> new TimestampField((String) args[0], (Map<String, Object>) args[1])
+            args -> new TimestampField((String) args[0])
```
This makes me wonder, this is fine as-is, but should we even bother having this class or specifying the name of the field if it's going to be hardcoded for the foreseeable future?

martijnvg replied:
I prefer to keep this class. I think that that the timestamp field will not stay hard coded for a long time.

dakrone reviewed (approved):
LGTM, thanks Martijn! I left some really minor comments

martijnvg committed:
only apply _timestamp meta field if index is created as part of a data stream or data stream rollover,
this fixes a docs test, where a regular index creation matches (logs-*) with a template with a data stream definition.

martijnvg committed:
Merge remote-tracking branch 'es/master' into fix_timestamp_field_to_@timestamp

martijnvg committed:
checkstyle

martijnvg committed:
fix ml integ tests

martijnvg committed:
Merge remote-tracking branch 'es/master' into fix_timestamp_field_to_@timestamp

martijnvg committed:
only include template with _timestamp field mapping if creation a backing index

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -902,6 +909,23 @@ public static String findV2Template(Metadata metadata, String indexName, boolean
             .ifPresent(mappings::add);
+
+        // Only include _timestamp mapping snippet if creating backing index.
+        if (indexName.startsWith(DataStream.BACKING_INDEX_PREFIX)) {
```
@dakrone I had to add this if statement otherwise the meta field mapper was going to be applied on each create index api call. In a docs test, the new logs-*-* composable index template was triggered by a regular create index api call and then the test failed, because the document being indexed had no timestamp field. 

I think only applying the meta field automatically makes sense for backing indices only and not when a user creates a new index via create index api and the index composable template matches?

See commits: a23d2e484594b5cbe788f3ef4bfcd40f97136f67, which then was superseded by: c78733204c12cf44b53026052aff9865d99ced03

martijnvg replied:
Note currently in master, the `_timestamp` field gets only applied if the create index request originates from create data stream api and data stream rollover (signalled via `CreateIndexClusterStateUpdateRequest#dataStreamName` field), but now this logic moved `MetadataIndexTemplateService.collectMappings(...)` method, which didn't have that context.

jtibshirani replied:
Would it be cleaner to pull this check out of `collectMappings` and into the caller? That way we could avoid the workaround here: https://github.com/elastic/elasticsearch/pull/59076/files#diff-115cd2f0fff9dd97acc95e1d29012a15R1075.

martijnvg committed:
one search and replace too many

martijnvg committed:
instead of changeing the current timestamp field, just add @timestamp and keep using current timestamp field

martijnvg committed:
Merge remote-tracking branch 'es/master' into fix_timestamp_field_to_@timestamp

martijnvg committed:
disable bwc tests

martijnvg committed:
iter

martijnvg committed:
iter2

martijnvg committed:
Merge remote-tracking branch 'es/master' into fix_timestamp_field_to_@timestamp

jtibshirani commented on server/src/main/java/org/elasticsearch/cluster/metadata/ComposableIndexTemplate.java:
```diff
@@ -267,6 +275,13 @@ public String getTimestampField() {
+        /**
+         * @return a mapping snippet for a backing index with `_timestamp` meta field mapper properly configured.
+         */
+        public Map<String, Object> getDataSteamMappingSnippet() {
```
Noticed this typo as I was trying to pull up the method: `DataSteam` -> `DataStream`.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/TimestampFieldMapper.java:
```diff
@@ -196,6 +196,10 @@ public void validate(DocumentFieldMappers lookup) {
         }
     }
 
+    public String getPath() {
```
Does `TimestampFieldMapper` need a `path` parameter now? We're not sure how we will eventually implement custom timestamp names, I can see an implementation where we don't use `path` here at all. So maybe we could simplify and just have an `enabled` boolean flag instead.

Depending on the implementation, we can always add `path` later and just default to `@timestamp` if it's not specified.

jtibshirani commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:
```diff
@@ -183,31 +169,16 @@ public static ComposableIndexTemplate lookupTemplateForDataStream(String dataStr
-        } else {
-            mappingPath = "properties." + fieldPath.replace(".", ".properties.");
+        if (timestampFieldName.equals(fieldMapper.getPath()) == false) {
+            throw new IllegalArgumentException("[_timestamp] meta field doesn't point to data stream timestamp field [" +
```
I think this error will be hard for end users to understand, since they shouldn't be aware of the `_timestamp` field. Maybe we won't even need this check, once we remove the ability to set `timestamp_field`, as discussed in https://github.com/elastic/elasticsearch/pull/59317#issuecomment-656236960?

martijnvg replied:
This is more of a sanity check. Kind of ensuring that we did apply the `_data_stream_timestamp` meta field mapper. If a user would run into this error then this would be a bug. Can tweak this check.

jtibshirani commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -1047,16 +1071,19 @@ private static void validateCompositeTemplate(final ClusterState state,
                     xContentRegistry, tempIndexService.newQueryShardContext(0, null, () -> 0L, null));
 
+                // triggers inclusion of _timestamp field and its validation:
+                String indexName = DataStream.BACKING_INDEX_PREFIX + temporaryIndexName;
```
Should we only add the `_timestamp` mapping if this template has a data stream block? Otherwise it could just be a normal template that happens to contain a field named `@timestamp`, and we attempt to validate that mapping.

jtibshirani reviewed (commented):
I left a few late comments!
